src/main/java/com/ibm/darpc/DaRPCMemPoolImpl.java

```diff
@@ -0,0 +1,202 @@
+				}
+			}
+			mrs = null;
+			for (String fileName : hugePageFiles) {
+				File f = new File(fileName);
+				f.delete();
+			}
+			hugePageFiles = null;
+		}
+	}
+
+	@Override
+	public ByteBuffer getBuffer(RdmaEndpoint endpoint) throws IOException, NoSuchElementException {
+		LinkedBlockingQueue<ByteBuffer> freeList = pdMap.get(endpoint.getPd());
+
+		if (freeList == null) {
+			synchronized(this) {
+				freeList = pdMap.get(endpoint.getPd());
+				if (freeList == null) {
+					freeList = new LinkedBlockingQueue<ByteBuffer>();
+					pdMap.put(endpoint.getPd(), freeList);
+				}
+			}
+		}
+
+		ByteBuffer r = freeList.poll();
+
+		if (r == null) {
+			synchronized(this) {
+				r = freeList.poll();
+				if (r == null) {
+					allocateHugePageBuffer(freeList, endpoint.getPd());
+				}
+				r = freeList.poll();
+				if (r == null) {
+					logger.error("Failed to allocate more buffers.");
+					throw new NoSuchElementException("Failed to allocate more buffers.");
+				}
+			}
+		}
+		r.clear();
+		return r;
+	}
+
+	@Override
+	public void freeBuffer(RdmaEndpoint endpoint, ByteBuffer buffer) {
+		LinkedBlockingQueue<ByteBuffer> freeList = pdMap.get(endpoint.getPd());
+		freeList.add(buffer);
+	}
+
+	@Override
+	public int getLKey(ByteBuffer buffer) throws IllegalArgumentException {
+		return memoryRegions.get(MemoryUtils.getAddress(buffer)).getLkey();
+	}
+
+	// allocate a buffer from hugepages
+	private void allocateHugePageBuffer(LinkedBlockingQueue<ByteBuffer> freeList, IbvPd pd) throws IOException {
+		int totalAllocationSize = allocationSize + alignmentSize;
+		if ((currentAllocationSize + totalAllocationSize) > allocationLimit) {
+			logger.error("Out of memory. Cannot allocate more buffers from hugepages. "
+					+ "allocationSize = " + allocationSize
+					+ ", alignmentSize = " + alignmentSize
+					+ ", currentAllocationSize = " + currentAllocationSize
+					+ ", allocationLimit = " + allocationLimit);
+			throw new IOException("Out of memory. Cannot allocate more buffers from hugepages."
+					+ "allocationSize = " + allocationSize
+					+ ", alignmentSize = " + alignmentSize
+					+ ", currentAllocationSize = " + currentAllocationSize
+					+ ", allocationLimit = " + allocationLimit);
+
+		}
+		String newFile = this.hugePagePath + hugePageFileName + System.currentTimeMillis();
+		RandomAccessFile randomFile = null;
+		try {
+			randomFile = new RandomAccessFile(newFile, "rw");
+		} catch (FileNotFoundException e) {
+			logger.error("Path " + newFile + " to huge page path/file cannot be accessed.");
+			throw e;
+		}
+		hugePageFiles.add(newFile);
+		try {
+			randomFile.setLength(totalAllocationSize);
+		} catch (IOException e) {
+			logger.error("Could not set allocation length of mapped random access file on huge page directory.");
+			logger.error("allocaiton size = " + allocationSize + " , alignment size =  " + alignmentSize);
+			logger.error("allocation size and alignment must be a multiple of the hugepage size.");
```

Comment:
multiple calls to the logger or append everything like above? I think we should be consistent.